### PR TITLE
cody web: provide suggestions in context selector

### PR DESCRIPTION
**Pull request by courier-new**
Closes https://github.com/sourcegraph/sourcegraph/issues/55363.

This PR provides suggestions to the context selector, so that it's less empty when you initially open the popover and it hopefully takes less work to find the repos you want to use Cody with.

### Screenshots
| Scenario | | Note |
|-----:|:-----------:|----|
| Popover first opened | image | The suggestions fill most of the popover window area and can be scrolled through. |
| When several suggestions have been added | image | Any repos added from suggestions move to populate the "Repositories" section. |
| Searching | image | The search UI remains unchanged. |
| With inferred repo/file | image | The suggestions fill in at the very bottom, distinct from the other sections.
| When all the suggestions have been added | image | The "Suggestions" section disappears, leaving only "Repositories". |

### Demo

The algorithm for suggesting these repos has a couple layers to it, so let me walk you through it via a demo!

_NB, the popover presented in the demos is taller than in the final version._

I'm starting as a brand new user who has never interacted with Sourcegraph or Cody before. This user can see 10 suggestions for repos when they open the context selector:

https://github.com/sourcegraph/sourcegraph/assets/8942601/cf9c46e4-2a07-4cb3-b838-66251dbd0ec8

The suggestions list takes on the same visual appearance as the currently selected repos list does. When the user selects a repo from suggestions, it moves from "Suggestions" to "Repositories".

Since this user has never interacted with Sourcegraph or Cody before, there's no history to make suggestions from. So instead, this initial list of suggestions is a combination of:
1. Any repositories from a hard-coded list of 10 of the most popular open source repositories on GitHub that are available on the instance.
2. The first 10 repositories with embeddings available on the instance, alphabetically.

In this demo, `sourcegraph/sourcegraph`, `facebook/react`, and `tensorflow/tensorflow` are the only 3 repos from the hard-coded list of most popular repos that I have indexed locally, which is why they appear first. The following 7 repos are the first 7 repos with embeddings available on my dev instance. On dotcom, a first-time user would see all 10 suggestions come from the hard-coded list of top open source repos. On a customer's instance that's only connected to their private code hosts, they'd see none of those.

In this demo, the user selects `sourcegraph-batch-changes/etcd` and `sourcegraph-batch-changes/src-cli` for their first conversation. Keep this in mind as we move forward!

Once the user goes to start a new chat with Cody, we see the new chat automatically preserves the last-used set of repos in context:

https://github.com/sourcegraph/sourcegraph/assets/8942601/a4f12226-eacb-4898-8334-1879f8da0cd6

The getting started banner reflects this. It's easy to reset or update the context. In this example, the user removes `sourcegraph-batch-changes/etcd` and adds `sourcegraph/sourcegraph`. When `etcd` is removed, we see it appear at the top of the suggestions list. This is because it's part of the user's transcript history, now, from the previous conversation thread. Repositories that the user has recently interacted with take precedence over the top 10 open source repos and the first 10 repos with embeddings.

Now the user goes to use code search, and we add repo search history into the mix:

https://github.com/sourcegraph/sourcegraph/assets/8942601/76f5d859-985c-40ba-bd87-1a42e7422b7a

We see here that the user first visits `github.com/sourcegraph/cody`, and then shortly after `github.com/sourcegraph/automation-testing`. Upon returning to the Chat UI and starting a new conversation, we can observe that:
- Before resetting the scope, these two repos are now at the top of the suggestions list, above `sourcegraph-batch-changes/etcd`, because they were interacted with more recently.
- After resetting the scope, the two repos that were removed are now at the top, then the two from search, then `etcd` from the earlier chat.

Suggestions from the user's recent activity are the ones that will be the most relevant to the user, so they always take priority over the others. Cody chat and code search are treated as equally important sources, so the repo suggestions that come from them are combined, de-duped, and sorted by last interaction time before being displayed.

At the end of this demo, we now have repos from 4 different sources integrated into the suggestions! Once the user has interacted with 10 different repos via Cody or search, they will be the only suggestions a user sees from then on out. For users who have already used code search and/or Cody, then, they'll see these suggestions immediately after we roll out these changes.

### Implementation Notes

We get the information on repos recently visited with search from the `useUserHistory` hook. I actually initially implemented my own take on this hook because I didn't find this one when I was searching for something like it. I only discovered it was there when I was probing local storage to test that my own version was working. 🤦‍♀️ Rookie mistake! While its name might suggest it captures more than just repos recently visited, I can confirm that's all it is actually tracking. For the most part, I was able to use this hook exactly as it is. I only changed the structure and contents stored by this hook in one meaningful way, which I will describe inline with the changes.

We're playing a bit fast-and-loose with the currently authenticated user in a couple places across the web app. This PR tries to clean that up in a couple small ways that were relevant to the context selector feature:
- Preferring to prop-drill `authenticatedUser` rather than read it from `window.context.currentUser`. This is because `window.context` is only loaded once, when the page initially loads, whereas `authenticatedUser` is updated by other app interactions. If the user is logged out, has their site admin status changed, or other things happen to them mid-session, for example, this wouldn't be reflected by `window.context.currentUser`.
- Re-indexing the user history in local storage by a user-differentiated storage key. Previously, all user history, regardless of who was currently logged in, was collected under the same key. Though the average user probably doesn't share their machine with other SG users, an admin may have several SG accounts for different purposes, so it's best to be able to distinguish this locally saved history by user ID. Plus it makes dev testing with different users a whole lot more sane.
- Re-indexing Cody chat transcript history in local storage by a user-differentiated storage key. This is the equivalent migration as is described for user history; previously, all transcripts were collected under the same key, regardless of who was logged in. They will now live under distinct storage keys.

## Test plan

- Walk through the steps of the demo as listed above, as a new user
- Tested interactions manually in standalone chat as well as sidebar chat
- Tested interactions manually in Web as well as App
- Check that CI is passing

### CODE REVIEW

File: client/web/src/components/useUserHistory.ts [hunk 1]
```diff
@@ -17,32 +18,53 @@ const LOCAL_STORAGE_KEY = 'user-history'
...
+    // entries and write them to the new user-differentiated key, and then delete the old
+    // key. When the method runs again in the future, it will thus be a no-op.
+    private migrateOldEntries(): void {
+        const oldJSON = this.storage.getItem(LOCAL_STORAGE_KEY)
+        if (!oldJSON) {
+            return
+        }
+        this.storage.setItem(this.storageKey(this.userID), oldJSON)
+        this.storage.removeItem(LOCAL_STORAGE_KEY)
+    }
```
Comment (courier-new) [inline client/web/src/components/useUserHistory.ts:55]: This is the only significant change I made to `useUserHistory`: namely, re-keying the history entries by user ID. To preserve the existing entries, I added this method for a one-time migration from `user-history` → `user-history:<USERID>` that just takes whatever we have today, assumes it belongs to the current user, and slaps it under the new local storage key.

File: client/web/src/components/useUserHistory.ts [hunk 2]
```diff
@@ -17,32 +18,53 @@ const LOCAL_STORAGE_KEY = 'user-history'
...
         for (let index = MAX_LOCAL_STORAGE_COUNT; index < entries.length; index++) {
             // Synchronize persisted entries with in-memory entries so that
             // reloading the page doesn't change which entries are available.
             this.deleteEntry(entries[index])
         }
     }
-    private loadEntries(): UserHistoryEntry[] {
-        return JSON.parse(this.storage.getItem(LOCAL_STORAGE_KEY) ?? '[]')
+    public loadEntries(): UserHistoryEntry[] {
+        return JSON.parse(this.storage.getItem(this.storageKey(this.userID)) ?? '[]')
```
Comment (courier-new) [inline client/web/src/components/useUserHistory.ts:67]: I made this public because the only other public method for listing entries just provides the repo names, and we actually want the `lastAccessed` timestamps, too, in order to interweave these with Cody chat history repos. 🙂

File: client/web/src/cody/components/ScopeSelector/ScopeSelector.module.scss [hunk 3]
```diff
@@ -55,15 +55,13 @@
 
 .repository-selector-content {
     width: 20rem;
-    height: 100%;
     max-width: 20rem;
-    min-height: 24rem;
-    max-height: 29rem;
+    height: 29rem;
```
Comment (courier-new) [inline client/web/src/cody/components/ScopeSelector/ScopeSelector.module.scss:59]: There's less layout shift if we just set a single height on the context selector popover like this. It also meant I could just `flex: 1` on the main child, which is nice.

File: client/web/src/cody/components/GettingStarted.tsx [hunk 4]
```diff
@@ -101,10 +104,10 @@ export const GettingStarted: React.FC<
             examples: [
                 {
                     label: 'Algorithms',
-                    text: 'How the QuickSort algorithm works with an implementation in Python?',
+                    text: 'Can you explain how the QuickSort algorithm works in Python?',
```
Comment (courier-new) [inline client/web/src/cody/components/GettingStarted.tsx:107]: I cleaned up the grammar/style in the "Getting Started" section just a tad while I was here!

File: client/web/src/cody/components/GettingStarted.tsx [hunk 5]
```diff
@@ -79,7 +82,7 @@ export const GettingStarted: React.FC<
             return {
                 title: `Great examples to start with${
                     scopeSelectorProps.scope.repositories.length === 1
-                        ? ` for ${scopeSelectorProps.scope.repositories[0]}`
+                        ? ` for ${scopeSelectorProps.scope.repositories[0].split('/').slice(-2).join('/')}`
```
Comment (courier-new) [inline client/web/src/cody/components/GettingStarted.tsx:85]: This just keeps the abbreviated `sourcegraph/sourcegraph` part of the repo name, dropping the code host from the start. Otherwise this can tend to get kinda long.

File: client/web/src/cody/components/CodeMirrorEditor.ts [hunk 6]
```diff
@@ -99,6 +99,10 @@ export class CodeMirrorEditor implements Editor {
         return null
     }
 
+    public getActiveTextEditorSelectionOrVisibleContent(): ActiveTextEditorSelection | null {
```
Comment (courier-new) [inline client/web/src/cody/components/CodeMirrorEditor.ts:102]: These changes, and the corresponding ones in `FileContentEditor.ts` and `RepoContainerEditor.ts`, are due to the `Editor` interface changing in the latest version of `@sourcegraph/cody-shared`. The change is unrelated to this PR, and the Class methods are only really relevant to VSCode (*shakes fist angrily at OOP*), but I have to implement them in order to upgrade that package version.

File: client/web/src/cody/useCodyChat.tsx [hunk 7]
```diff
@@ -135,6 +145,7 @@ export const useCodyChat = ({
             customHeaders: window.context.xhrHeaders,
             debugEnable: false,
             needsEmailVerification: isEmailVerificationNeededForCody(),
+            experimentalLocalSymbols: false,
```
Comment (courier-new) [inline client/web/src/cody/useCodyChat.tsx:148]: This is also an API change from `@sourcegraph/cody-shared` v0.0.6.

**Review by toolmantim — APPROVED** [review]
I gave this a quick review and it looks solid. There's some design suggestions that I left in the Figma a while back (e.g. `s/repos/repositories`) that I'll leave out of this PR and let @danielmarquespt revisit. Approving from a design POV ✅

There's a few non-blocking small things I spotted:

Add a bit more padding (`pt-2) to the "Repositories" and "Suggestions" separator headings:

Screenshot 2023-08-23 at 11 00 05 am

Fix the vertical alignment of the radio and it's label (`align-items-center` instead of `align-items-baseline`)

And let's update this description:

> Some repos are not indexed for Cody. This may affect the quality of the answers. Learn more about this [in the docs](https://sourcegraph.test:3443/help/cody/explanations/code_graph_context#embeddings).

to:

> 4 of 6 selected repositories are not indexed for Cody, and are missing embeddings. This may affect the quality of the answers. To enable indexing, see the [embeddings documentation](https://sourcegraph.test:3443/help/cody/explanations/code_graph_context#embeddings).

For a single repo:

> The selected repository is not indexed for Cody, and is missing embeddings. This may affect the quality of the answers. To enable indexing, see the [embeddings documentation](https://sourcegraph.test:3443/help/cody/explanations/code_graph_context#embeddings).

(and could we make the `<a>` match the orange text color, with an underline?)

Thanks @courier-new!

**Comment by toolmantim** [comment]
Oh, and one more tiny non-blocking request/question:
- Can we remove the word "Great" from the examples heading?
- Have we confirmed that these do in fact return great results? When I clicked "Can you describe the overall structure of this repo?" it replies "Unfortunately I do not have enough context […]" which isn't a great getting started experience.

File: client/web/src/cody/components/GettingStarted.tsx [hunk 8]
```diff
@@ -193,16 +196,17 @@ export const GettingStarted: React.FC<
...
                                 <Text size="small" className={classNames('text-muted', styles.hintTitle)}>
-                                    Why context is important?
+                                    Why is context important?
                                 </Text>
                                 <Text size="small" className={classNames('text-muted', styles.hintText)}>
-                                    Without providing a specific repo as context, Cody won’t be able to answer with
-                                    relevant knowledge about your project.
+                                    Without providing relevant repo(s) for context, Cody won't be able to answer
+                                    questions specific to your project.
                                 </Text>
```
[the remark this answers is not in the document]
Comment (courier-new) [reply client/web/src/cody/components/GettingStarted.tsx]: [Done!](https://github.com/sourcegraph/sourcegraph/pull/56039/commits/e44d7858cf845b7abd0cd04e4e7ebda85a91f703)

**Comment by courier-new** [comment]
Thanks for the feedback, Tim! I've implemented all of your suggestions. 🙂

**Comment by taras-yemets** [comment]
@courier-new, these prompts were suggested in the [original design](https://www.figma.com/file/al8wd8wOIrxhgjsu9IcsK7/Cody-Web-%5BMAIN%5D?node-id=2214:118&mode=dev). We are going to reiterate on them once we have some initial feedback.

> Have we confirmed that these do in fact return great results? When I clicked "Can you describe the overall structure of this repo?" it replies "Unfortunately I do not have enough context […]" which isn't a great getting started experience.

@toolmantim, I agree the experience could be better. This response may be because of not indexed repos in the scope selector. It should be better for indexed repos https://github.com/sourcegraph/sourcegraph/pull/55548#discussion_r1292953666. We may want to use different sets of example prompts for one repo selected / multiple repos selected / all repos indexed / some repos not indexed / all repos not indexed use cases.

**Comment by toolmantim** [comment]
@courier-new 🎉 Looks good!

@taras-yemets I like you're idea of changing it depending on embeddings status — just to ensure we don't have a bad first impression.

I did just spot another thing, once I tested those examples some more: we should probably only change the examples once they've selected a repository, otherwise they'll get unhappy response:

https://github.com/sourcegraph/sourcegraph/assets/153/80670c64-57b2-47c7-a9f1-1aa39110b708

This could be a bug just on my machine, and maybe unrelated to this PR, but if you choose a suggestion, then "Clear all chats", and then click around in the two elements (?) the message re-appears (and sometimes with a typewriter closed error):

https://github.com/sourcegraph/sourcegraph/assets/153/c9807b6e-d5e5-4b90-8784-d1a5cde64815

**Comment by taras-yemets** [comment]
>This could be a bug just on my machine, and maybe unrelated to this PR, but if you choose a suggestion, then "Clear all chats", and then click around in the two elements (?) the message re-appears (and sometimes with a typewriter closed error)

Thanks for reporting, @toolmantim! I wasn't able to reproduce it either on dotcom or locally from this branch 🤔

**Comment by taras-yemets** [comment]
> I did just spot another thing, once I tested those examples some more: we should probably only change the examples once they've selected a repository, otherwise they'll get unhappy response.

Agree! Will fix it! Thank you!

**Review by thenamankumar — APPROVED** [review]
LGTM

**Comment by courier-new** [comment]
Please feel free to merge this on my behalf, as my parting contribution, then. <3 It was great to have the opportunity to work with you all, however brief it was.
